src/twisted/conch/ssh/transport.py

```diff
@@ -516,6 +515,99 @@ class SSHTransportBase(protocol.Protocol):
+        Then call addSupportedKexes to populate the list.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        # Clear out the currently supported key exchanges.
+        _kex._kexAlgorithms.clear()
+
+        # Set the new exchanges.
+        self.addSupportedKexes(kexes)
+
+
+    def addSupportedKexes(self, kexes):
+        """
+        Append the list of supported key exchanges.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        for k in kexes:
+            if k.startswith(b'ecdh') and kexes[k] is None:
+                size = int(k[-3:])
+
+                # Add the new kex to the dictionary.
+                if size <= 256:
+                    _kex._kexAlgorithms[k] = _kex._ECDH256()
+                elif size <= 384:
+                    _kex._kexAlgorithms[k] = _kex._ECDH384()
+                else:
+                    _kex._kexAlgorithms[k] = _kex._ECDH512()
+            elif kexes[k] is None:
+                _kex._kexAlgorithms[k] = _kex._baseKexAlgorithms[k]
+            else:
+                _kex._kexAlgorithms[k] = kexes[k]
+
+        # Reload
+        del self.supportedKeyExchanges[:]
+        for kex in _kex.getSupportedKeyExchanges():
+            self.supportedKeyExchanges.append(kex)
+
+
+    def setSupportedPublicKeys(self, publicKeys):
+        """
+         Clear out the list of supported public keys
+         Then call addSupportePublicKeys to populate the list.
+
+         @type publicKeys: L{dict}
+         @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        del self.supportedPublicKeys[:]
+        self.addSupportePublicKeys(publicKeys)
+
+
+    def addSupportePublicKeys(self, publicKeys, sort=True):
+        """
+        Populate the supported public key list.
+
+        @type publicKeys: L{dict}
+        @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        for k in publicKeys:
+            if k not in self.supportedPublicKeys:
+                self.supportedPublicKeys.append(k)
+
+            if k.startswith(b'ecdsa'):
+                curve = publicKeys[k]
+                keys._curveTable[k] = curve[0]
+
+                keys._secToNist[curve[1]] = k
+
+        # Sort the keys so the strongest ones are preferred?
+        if sort:
+            # Reverse sort to bring the stronger keys up first.
+            self.supportedPublicKeys.sort(reverse=True)
+
+            # If 'ssh-rsa' or 'ssh-dss' are present, remove and add
+            # to make sure they're in the right order
+            if b'ssh-rsa' in self.supportedPublicKeys:
```

Comment:
I'm also unsure about the sort order!  This moves RSA and DSA key types to the end of the list, but what about the relative order of elliptic curve key types?

[`openssh` allows users to specify the complete list, to remove items from the default, or to _append_ items to it](https://man.openbsd.org/sshd_config#PubkeyAcceptedKeyTypes).  That seems like an acceptable API.  This could also be applied to changing the supported key exchange algorithms.